rfd/0099-link-bundled-tsh.md

```diff
@@ -0,0 +1,127 @@
+commands](https://code.visualstudio.com/docs/setup/mac#_launching-from-the-command-line) for
+installing and uninstalling the CLI tool. The installation command executes an AppleScript with
+administrator privileges which symlinks `code` to `/usr/local/bin/code`.
+
+* https://github.com/microsoft/vscode/blob/1.74.0/src/vs/workbench/electron-sandbox/actions/installActions.ts
+* https://github.com/microsoft/vscode/blob/1.74.0/src/vs/platform/native/electron-main/nativeHostMainService.ts#L261
+
+### How electron-builder does it
+
+#### Linux
+
+Linux is the only platform where during the installation of a package, electron-builder
+automatically symlinks the binary to `/usr/bin`.
+
+Unlike VSCode, electron-builder doesn't have special templates for any particular package format.
+Instead it uses [fpm](https://fpm.readthedocs.io) which handles many different formats.
+
+The binary is linked through a post install and post removal scripts. This means that rpm doesn't
+recognize that the binary comes from the specific package.
+
+* https://github.com/electron-userland/electron-builder/blob/v24.0.0-alpha.5/packages/app-builder-lib/templates/linux/after-install.tpl
+* https://github.com/electron-userland/electron-builder/blob/v24.0.0-alpha.5/packages/app-builder-lib/templates/linux/after-remove.tpl
+
+### How Connect can do it
+
+#### Linux
+
+We can automatically link tsh using the same method that electron-builder uses. Unfortunately,
+electron-builder doesn't seem to provide a way to extend the post install script in any way. This
+most likely means that we'll have to copy the default script and keep it in sync when updating
+electron-builder.
+
+Instead of symlinking the binary to `/usr/bin/tsh`, we can symlink it to `/usr/local/bin/tsh`. This
+is where our Linux teleport package places our binaries. This way if someone installs Connect after
+installing teleport, we'll overwrite just a single symlink instead of the user ending up with two
+different symlinks for tsh.
+
+When adding the custom script, we might want to change the location of the `teleport-connect`
+symlink to `/usr/local/bin` as well.
+
+#### Windows
+
+We use NSIS to build the Windows installer. electron-builder offers a way to extend the scripts
+executed by the installer by [defining custom
+macros](https://www.electron.build/configuration/nsis#custom-nsis-script). NSIS has [built-in
+utilities to manipulate the `Path` env var](https://nsis.sourceforge.io/Path_Manipulation), however
+at the top of that page there's a big red banner telling you to use [the EnVar
+plugin](https://nsis.sourceforge.io/EnVar_plug-in) for path manipulation instead due to length
+limitations in the built-in utilities.
+
+We will have to figure out how to use that plugin but once that's done, we should be able to simply
+add the bin folder to the path, just as VSCode does.
+
+#### macOS
+
+For now we've decided to introduce a custom `tsh install` command to the suggested commands in the
```

Comment:
This is a little confusing since it will look like a real tsh command, but it's fine for now.

What would we need to do in order to make the command bar more like a VS Code style command palette rather than just a tsh command executor?

Reply:
> What would we need to do in order to make the command bar more like a VS Code style command palette rather than just a tsh command executor?

The first version used to work like the command palette, the first think I worked on was ripping it out and replacing it with autocomplete hehe.

The command palette version was fundamentally incompatible with autocomplete AFAIK. But we could do the same thing VSCode does: make it perform autocomplete by default but if you start the input with `>` then it becomes we replace it with the initial implementation and it becomes a command palette. I think we even considered that at one point.